src/services/firm_quote_validator.ts

```diff
@@ -0,0 +1,147 @@
+            if (makerTokenBalanceForMaker === undefined) {
+                makerAddressesToAddToCacheSet.add(quote.makerAddress);
+                return quote.takerAssetAmount;
+            }
+
+            // Order is fully fillable, because Maker has 100% of the assets
+            if (makerTokenBalanceForMaker.gte(quote.makerAssetAmount)) {
+                return quote.takerAssetAmount;
+            }
+
+            // Order is empty, return zero
+            if (quote.makerAssetAmount.lte(0)) {
+                return BIG_NUMBER_ZERO;
+            }
+
+            // Order is partially fillable, because Maker has a fraction of the assets
+            const partialFillableAmount = makerTokenBalanceForMaker.times(quote.takerAssetAmount).div(quote.makerAssetAmount).integerValue(BigNumber.ROUND_DOWN);
+            if (!partialFillableAmount.isFinite()) {
+                logger.error(`Calculated maker token balance is infinite, which caused the partialFillableAmount to be infinite. This should never happen`);
+                return BIG_NUMBER_ZERO;
+            }
+            return partialFillableAmount;
+        });
+        const makerAddressesToAddToCache = Array.from(makerAddressesToAddToCacheSet);
+        return {
+            makerAddressesToAddToCache,
+            takerFillableAmounts,
+        };
+    }
+
+    constructor(chainCacheRepository: Repository<MakerBalanceChainCacheEntity>, cacheExpiryThresholdMs: number = THRESHOLD_CACHE_EXPIRED_MS) {
+        this._chainCacheRepository = chainCacheRepository;
+        this._cacheExpiryThresholdMs = cacheExpiryThresholdMs;
+    }
+
+    // tslint:disable-next-line: prefer-function-over-method
+    public async getRFQTTakerFillableAmounts(quotes: SignedOrder[]): Promise<BigNumber[]> {
+        // TODO: Handle error on query
+
+        // Ensure that all quotes have the same exact maker token.
+        const makerTokenAddressesSet = new Set(
+            quotes.map(quote => {
+                const decodedAssetData = assetDataUtils.decodeAssetDataOrThrow(quote.makerAssetData) as ERC20AssetData;
+                return decodedAssetData.tokenAddress;
+            })
+        );
+        const makerTokenAddresses = Array.from(makerTokenAddressesSet);
+        if (makerTokenAddresses.length !== 1) {
+            logger.error(`Quotes array was empty or found multiple maker token addresses within one single RFQ batch: ${JSON.stringify(makerTokenAddresses)}. Rejecting the batch`);
+            return quotes.map(_quote => BIG_NUMBER_ZERO);
+        }
+        const makerTokenAddress = makerTokenAddresses[0];
+
+        // Fetch balances and create a lookup table
+        const makerLookup: {[key: string]: BigNumber} = {};
+        const makerAddresses = Array.from(new Set(quotes.map(quote => quote.makerAddress)));
+        const cacheResults = await this._chainCacheRepository.find({
+            where: [{
+                tokenAddress: makerTokenAddresses[0],
```

Comment:
we already pull out the first element above and name it `makerTokenAddress`

Reply:
done